File: LeetCode_List/147Insertion_Sort_List(fianl).py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy=ListNode(0)

        while head:
        	iter=dummy;
        	while iter.next and iter.next.val<head.val:
        		iter=iter.next
        	tmpNext=head.next
        	head.next=iter.next
        	iter.next=head
        	head=tmpNext

        return dummy.next

File: LeetCode_List/147Insertion_Sort_List(fianl).cpp
// 更为简练 思想是一样的
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode *dummy=new ListNode(0);

        while(head) {
            ListNode *iter=dummy;

            while(iter->next && iter->next->val < head->val)  { iter=iter->next; }

            ListNode *next=head->next;
            head->next=iter->next;
            iter->next=head;
            head=next;
        }

        return dummy->next;

    }
};